br/pkg/storage/gcs_extra.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	client := resty.New()
+	resp, err := client.R().SetBody(bodyXML).Post(u)
+	if err != nil {
+		return errors.Errorf("POST request failed: %s", err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return errors.Errorf("POST request returned non-OK status: %d", resp.StatusCode())
+	}
+
+	return nil
+}
+
+func buildCompleteMultipartUploadXML(parts map[int]objectPart) (string, error) {
+	xmlStr := strings.Builder{}
+	encoder := xml.NewEncoder(&xmlStr)
+	encoder.Indent("", "  ") // Indent with 2 spaces.
+
+	upload := struct {
+		XMLName xml.Name     `xml:"CompleteMultipartUpload"`
+		Parts   []objectPart `xml:"Part"`
+	}{}
+	// Order parts in the XML request.
+	upload.Parts = make([]objectPart, 0, len(parts))
+	for partNum := 1; partNum <= len(parts); partNum++ {
+		part, ok := parts[partNum]
+		if !ok {
+			return "", errors.Errorf("part %v not contained in parts", partNum)
+		}
+		upload.Parts = append(upload.Parts, part)
+	}
+
+	if err := encoder.Encode(upload); err != nil {
+		return "", err
+	}
+	return xmlStr.String(), nil
+}
+
+type objectPart struct {
+	PartNumber int    `xml:"PartNumber"`
+	ETag       string `xml:"ETag"`
+}
+
+type uploadWorker struct {
+	cli   *storage.Client
+	uri   string
+	index int // the index of the worker
+
+	bucketName, uploadID string
+
+	buffer            []byte // buffer capacity is equal to partSize
+	offset            int
+	partSize          int // 5 MiB <= partSize <= 5 GiB
+	currentPartNumber int
+	parts             map[int]objectPart
+
+	idleWorkerPool chan *uploadWorker
+
+	err    error // the error got from the last upload, will retry this part
+	tryCnt int
+}
+
+func newUploadWorker(cli *storage.Client, uri string, bucketName, uploadID string, index, partSize int, idleWorkerPool chan *uploadWorker) *uploadWorker {
+	return &uploadWorker{
+		uri:            uri,
+		cli:            cli,
+		index:          index,
+		bucketName:     bucketName,
+		uploadID:       uploadID,
+		buffer:         make([]byte, partSize),
+		partSize:       partSize,
+		parts:          make(map[int]objectPart),
+		idleWorkerPool: idleWorkerPool,
+	}
+}
+
+func (uw *uploadWorker) bufferData(partNumber int, p []byte) int {
+	uw.currentPartNumber = partNumber
+	n := copy(uw.buffer[uw.offset:], p)
+	uw.offset += n
+	return n
+}
+
+func (uw *uploadWorker) full() bool {
+	return uw.offset >= uw.partSize // if true, this worker is ready to upload the full part
+}
+
+func (uw *uploadWorker) uploadPartAsync() error {
+	if uw.err != nil && uw.tryCnt >= uploadPartTryCnt {
+		return errors.Errorf("failed to upload part %v too many times, err: %s", uw.index, uw.err.Error())
+	}
+	go func() {
+		err := uw.uploadPart(uw.currentPartNumber, uw.buffer[:uw.offset]) // timeout in few seconds
+		if err != nil {
+			uw.err = err
+			uw.tryCnt++
+			log.Warn(fmt.Sprintf("upload worker %v failed to upload part %v for object %v, tryCnt: %v, err: %s", uw.index, uw.currentPartNumber, uw.uri, uw.tryCnt, err.Error()))
```

Comment:
also change this logger. And we should not put the worker into `idleWorkerPool`? Otherwise currentPartNumber will be overwritten